wasm/src/graph.rs

```diff
@@ -1212,49 +1350,55 @@ impl NoteGraph {
         }
     }
 
-    pub fn int_safe_add_edge(&mut self, from_path: &String, to_path: &String, edge_type: &String) {
-        let from = self.node_hash.get(from_path);
-        let to = self.node_hash.get(to_path);
+    pub fn int_safe_add_edge(
+        &mut self,
+        source_path: &String,
+        target_path: &String,
+        edge_type: &String,
+        edge_source: &String,
+    ) {
+        let source = self.node_hash.get(source_path);
+        let target = self.node_hash.get(target_path);
 
-        let from_index: NodeIndex<u32>;
-        let to_index: NodeIndex<u32>;
+        let source_index: NodeIndex<u32>;
+        let target_index: NodeIndex<u32>;
         let mut add_from_to_hash = false;
         let mut add_to_to_hash = false;
 
-        if from.is_none() {
-            from_index = self
+        if source.is_none() {
+            source_index = self
                 .graph
-                .add_node(NodeData::new_unresolved(from_path.clone()));
+                .add_node(NodeData::new_unresolved(source_path.clone()));
             add_from_to_hash = true;
         } else {
-            from_index = from.unwrap().clone();
+            source_index = source.unwrap().clone();
         }
 
-        if to.is_none() {
-            if from_path == to_path {
-                to_index = from_index;
+        if target.is_none() {
+            if source_path == target_path {
+                target_index = source_index;
             } else {
-                to_index = self
+                target_index = self
                     .graph
-                    .add_node(NodeData::new_unresolved(to_path.clone()));
+                    .add_node(NodeData::new_unresolved(target_path.clone()));
                 add_to_to_hash = true;
             }
         } else {
-            to_index = to.unwrap().clone();
+            target_index = target.unwrap().clone();
         }
 
         if add_from_to_hash {
-            self.node_hash.insert(from_path.clone(), from_index);
+            self.node_hash.insert(source_path.clone(), source_index);
         }
 
         if add_to_to_hash {
-            self.node_hash.insert(from_path.clone(), to_index);
+            self.node_hash.insert(source_path.clone(), target_index);
         }
```

Comment:
Curious why don't you just add it to the hash where you set the flag to true, instead?

Reply:
Borrow checking rules. `source` and `target` point into the hashmap. If I mutate the hashmap, the references to `source` and `target` might become invalid. So I can't use `source` and `target` anymore after mutating the hashmap. So I have to make the insertions after using `source` and `target`.